### PR TITLE
ci: add job to generate preview of differences to generated code

### PR DESCRIPTION
**Pull request by aajtodd**
## Issue \#

N/A

## Description of changes

Taken from Rust, this PR adds a CI job to generate a preview of differences to the generated code for all protocols. It works by generating the code and checking it into a branch `__generated-<branch-name` and diffs it against `__generated-main`. The main branch is refreshed on every PR merge. The generated branches are deleted when PR is closed.

The CI job will comment on the PR when a diff is ready and provide a link to compare the changes. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/scripts/mk-generated.sh [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# This script outputs the name of the new branch to stdout
+#
+# Below is the set of services that are generated for codegen preview
+# These are carefully selected to exercise every Smithy protocol.
+#  - @awsJson1_0: dynamodb
+#  - @awsJson1_1: config
+#  - @awsQuery: sts
+#  - @ec2Query: ec2
+#  - @restJson1: polly
+#  - @restXml: s3
```
Comment (ianbotsf) [inline .github/scripts/mk-generated.sh:19]: **Comment**: Nice.

**Review by ianbotsf — APPROVED** [review]
Looks helpful. I'll be eager to see an example once we have some new codegen changes.

File: .github/scripts/mk-generated.sh [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    base_branch=__generated-$base_branch
+  else
+    base_branch=__generated__
+  fi
+  echo "Loaded branch from GitHub: $gh_branch ($GITHUB_HEAD_REF). Base branch: $base_branch"
+  current_branch="${gh_branch:-$(git rev-parse --abbrev-ref HEAD)}"
+  echo "Current branch resolved to: $current_branch"
+  gen_branch="__generated-$current_branch"
+  repo_root=$(git rev-parse --show-toplevel)
+  cd "$repo_root" && ./gradlew :codegen:sdk:bootstrap -Paws.services=+dynamodb,+config,+sts,+ec2,+polly,+s3
```
Comment (kggilmer) [inline .github/scripts/mk-generated.sh:38]: suggestion

Maybe pull the service list spec out into a param passed to the script

File: .github/workflows/codegen-preview.yml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      # this is a load-bearing branch filter: if this isn't here, you may
+      # end up generating diffs for __generated-* branches which would lead to infinite recursion...
+      - main
+  pull_request:
+    types:
+      - opened
+      - reopened
+      - closed
+      - synchronize
+env:
```
Comment (kggilmer) [inline .github/workflows/codegen-preview.yml:16]: suggestion

pass the services generated in the shell script as a env variable here

File: builder.json [hunk 4]
```diff
@@ -38,6 +38,11 @@
             "!test_steps": [
                 "{gradlew} test allTests"
             ]
+        },
```
Comment (kggilmer) [inline builder.json:41]: question

for my understanding, what's this doing?

**Review by kggilmer — APPROVED** [review]

File: builder.json (hunk 4 above)
Comment (aajtodd) [reply builder.json:41]: it's defining a variant that disables any build or test step. We don't actually need to build the runtime and `smithy-aws-kotlin-codegen` is built automatically as needed by the bootstrap task called by `mk-generated.sh`. It's just an optimization to build only what's needed.

**Pull request merged**